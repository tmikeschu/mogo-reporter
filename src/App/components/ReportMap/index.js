import React, { PureComponent } from "react";
import {
  withScriptjs,
  withGoogleMap,
  GoogleMap,
  Marker,
} from "react-google-maps";

const AsyncMap = withScriptjs(
  withGoogleMap(props => (
    /* eslint-disable no-undef */
    // allow global google call
    <GoogleMap {...props.mapDefaults} ref={props.onMapMounted}>
      <Marker position={{ ...props.mapDefaults.defaultCenter }} />
    </GoogleMap>
  ))
);

class ReportMap extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      defaultCenter: {
        lat: 42.3603416,
        lng: -83.0801975,
      },
      defaultZoom: 11,
    };
  }

  componentDidMount() {
    /* eslint-disable no-undef */
    // allow navigator call
    if (navigator && navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(this.setCurrentLocation);
    }
  }

  componentDidUpdate(_, { defaultCenter: prevCenter }) {
    const { defaultCenter: currentCenter } = this.state;
    if (prevCenter !== currentCenter) {
      this.map.panTo({ ...currentCenter });
    }
  }

  setCurrentLocation = pos => {
    const { coords: { latitude: lat, longitude: lng } = {} } = pos;
    this.setState({ defaultCenter: { lat, lng } });
  };

  onMapMounted = ref => {
    this.map = ref;
    this.forceUpdate();
  };

  render() {
    return (
      <AsyncMap
        googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyDttFtNppkWvWarNnGNvJsCEynQBVfO2K4&v=3.exp&libraries=geometry,drawing,places"
        loadingElement={<div style={{ height: `100%` }} />}
        containerElement={<div style={{ height: `100vh` }} />}
        mapElement={<div style={{ height: `100%` }} />}
        mapDefaults={{ ...this.state }}
        onMapMounted={this.onMapMounted}
        {...this.props}
      />
    );
  }
}

export default ReportMap;
