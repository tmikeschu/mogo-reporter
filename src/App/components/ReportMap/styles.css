.ReportMap {
  height: 100%;
  position: absolute;
  width: 0;
  overflow: hidden;
  transition: width 500ms;
}

.ReportMap.visible {
  width: 100%;
}
