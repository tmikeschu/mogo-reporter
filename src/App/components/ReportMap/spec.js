import React from "react";
import { mount } from "enzyme";
import ReportMap from "./";

describe("<ReportMap />", () => {
  const wrapper = shallow(<ReportMap />);
  const container = wrapper.instance();

  it("renders without crashing", () => {
    expect(wrapper).toBeTruthy();
    expect(container).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("mounts without crashing", () => {
    const wrapper = mount(<ReportMap />);
    expect(wrapper).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  describe("componentDidMount", () => {
    it("sends a callback to the navigator.geolocation.getCurrentPosition", () => {
      navigator.geolocation = {
        getCurrentPosition: jest.fn(),
      };
      const mock = navigator.geolocation.getCurrentPosition;
      container.componentDidMount();
      expect(mock).toHaveBeenCalledWith(container.setCurrentLocation);
      mock.mockReset();
    });
  });

  describe("setCurrentLocation", () => {
    it("sets state to the position object coords", () => {
      const position = { coords: { latitude: 1, longitude: 2 } };
      expect(container.state.defaultCenter.lat).not.toEqual(1);
      expect(container.state.defaultCenter.lng).not.toEqual(2);

      container.setCurrentLocation(position);

      expect(container.state.defaultCenter.lat).toEqual(1);
      expect(container.state.defaultCenter.lng).toEqual(2);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("componentDidUpdate", () => {
    it("calls panTo on map if states are different", () => {
      const position = { defaultCenter: { lat: 1, lng: 2 } };
      container.map = { panTo: jest.fn() };
      const mock = container.map.panTo;
      container.componentDidUpdate({}, position);
      expect(mock).toHaveBeenCalledWith(position.defaultCenter);

      delete container.map;
    });

    it("doesnt call panTo on map if states are the same", () => {
      const position = container.state;
      container.map = { panTo: jest.fn() };
      const mock = container.map.panTo;
      container.componentDidUpdate({}, position);
      expect(mock).not.toHaveBeenCalled();

      delete container.map;
    });
  });

  describe("onMapMounter", () => {
    it("sets this.map to the ref", () => {
      expect(container.map).toBeUndefined();
      container.onMapMounted("MAP");
      expect(container.map).toEqual("MAP");
    });

    it("calls forceUpdate", () => {
      const restore = container.forceUpdate;
      container.forceUpdate = jest.fn();
      const mock = container.forceUpdate;
      container.onMapMounted("MAP");
      expect(mock).toHaveBeenCalled();
      container.forceUpdate = restore;
    });
  });
});
