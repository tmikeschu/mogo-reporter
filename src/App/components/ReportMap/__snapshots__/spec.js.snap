// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReportMap /> mounts without crashing 1`] = `
<ReportMap>
  <withScriptjs(withGoogleMap(Component))
    containerElement={
      <div
        className="ReportMap"
      />
    }
    googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyDttFtNppkWvWarNnGNvJsCEynQBVfO2K4&v=3.exp&libraries=geometry,drawing,places"
    loadingElement={
      <div
        style={
          Object {
            "height": "100%",
          }
        }
      />
    }
    mapDefaults={
      Object {
        "defaultCenter": Object {
          "lat": 42.3603416,
          "lng": -83.0801975,
        },
        "defaultZoom": 11,
      }
    }
    mapElement={
      <div
        style={
          Object {
            "height": "100%",
          }
        }
      />
    }
    onMapMounted={[Function]}
  >
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  </withScriptjs(withGoogleMap(Component))>
</ReportMap>
`;

exports[`<ReportMap /> renders without crashing 1`] = `
<withScriptjs(withGoogleMap(Component))
  containerElement={
    <div
      className="ReportMap"
    />
  }
  googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyDttFtNppkWvWarNnGNvJsCEynQBVfO2K4&v=3.exp&libraries=geometry,drawing,places"
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapDefaults={
    Object {
      "defaultCenter": Object {
        "lat": 42.3603416,
        "lng": -83.0801975,
      },
      "defaultZoom": 11,
    }
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  onMapMounted={[Function]}
/>
`;

exports[`<ReportMap /> setCurrentLocation sets state to the position object coords 1`] = `
<withScriptjs(withGoogleMap(Component))
  containerElement={
    <div
      className="ReportMap"
    />
  }
  googleMapURL="https://maps.googleapis.com/maps/api/js?key=AIzaSyDttFtNppkWvWarNnGNvJsCEynQBVfO2K4&v=3.exp&libraries=geometry,drawing,places"
  loadingElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  mapDefaults={
    Object {
      "defaultCenter": Object {
        "lat": 42.3603416,
        "lng": -83.0801975,
      },
      "defaultZoom": 11,
    }
  }
  mapElement={
    <div
      style={
        Object {
          "height": "100%",
        }
      }
    />
  }
  onMapMounted={[Function]}
/>
`;
